types/express-serve-static-core/ts4.0/express-serve-static-core-tests.ts

```diff
@@ -0,0 +1,161 @@
+});
+
+// Query will be defaulted to Query type - under route
+app.route('/:foo').get(req => {
+    req.query; // $ExpectType ParsedQs
+});
+
+// Next can receive a Error parameter to delegate to Error handler
+app.get('/nexterr', (req, res, next) => {
+    next(new Error('dummy')); // $ExpectType void
+});
+
+// Next can receive a 'router' parameter to fall back to next router
+app.get('/nextrouter', (req, res, next) => {
+    next('router'); // $ExpectType void
+});
+
+// Default types
+app.post('/', (req, res) => {
+    req.params[0]; // $ExpectType string
+
+    req.body; // $ExpectType any
+    res.send('ok'); // $ExpectType Response<any, Record<string, any>, number>
+});
+
+// Default types - under route
+app.route('/').post((req, res) => {
+    req.params[0]; // $ExpectType string
+
+    req.body; // $ExpectType any
+    res.send('ok'); // $ExpectType Response<any, Record<string, any>, number>
+});
+
+// No params, only response body type
+app.get<never, { foo: string }>('/', (req, res) => {
+    req.params.baz; // $ExpectError
+
+    res.send({ foo: 'ok' }); // $ExpectType Response<{ foo: string; }, Record<string, any>, number>
+    req.body; // $ExpectType any
+});
+
+// No params, only response body type - under route
+app.route('/').get<never, { foo: string }>((req, res) => {
+    req.params.baz; // $ExpectError
+
+    res.send({ foo: 'ok' }); // $ExpectType Response<{ foo: string; }, Record<string, any>, number>
+    req.body; // $ExpectType any
+});
+
+// No params, request body type and response body type
+app.post<never, { foo: string }, { bar: number }>('/', (req, res) => {
+    req.params.baz; // $ExpectError
+
+    res.send({ foo: 'ok' }); // $ExpectType Response<{ foo: string; }, Record<string, any>, number>
+    req.body.bar; // $ExpectType number
+
+    res.json({ baz: 'fail' }); // $ExpectError
+    req.body.baz; // $ExpectError
+});
+
+// No params, request body type and response body type - under route
+app.route('/').post<never, { foo: string }, { bar: number }>((req, res) => {
+    req.params.baz; // $ExpectError
+
+    res.send({ foo: 'ok' }); // $ExpectType Response<{ foo: string; }, Record<string, any>, number>
+    req.body.bar; // $ExpectType number
+
+    res.json({ baz: 'fail' }); // $ExpectError
+    req.body.baz; // $ExpectError
+});
+
+app.engine('ntl', (_filePath, _options, callback) => {
+    callback(new Error('not found.'));
+});
+
+// Status test
+{
+    type E = express.Response<unknown, any, 'abc'>; // $ExpectError
+    type B = express.Response<unknown, any, 123>;
+    type C = Parameters<B['status']>[0]; // $ExpectType 123
+    type D = Parameters<B['sendStatus']>[0]; // $ExpectType 123
+}
+
+// Locals can be a custom type
+app.get<{}, any, any, {}, { foo: boolean }>('/locals', (req, res, next) => {
+    res.locals.foo; // $ExpectType boolean
+    res.locals.bar; // $ExpectError
+    res.send({ foo: 'ok' }); // $ExpectType Response<any, { foo: boolean; }, number>
+});
```

Comment:
The Express docs have a few examples of tricky routes: https://expressjs.com/en/guide/routing.html#route-parameters

It would be a good idea to add them to the tests here to make sure they work as expected:
- `/flights/:from-:to`
- `/plantae/:genus.:species`
- `/user/:userId(\d+)`

While it's not explicitly mentioned on that page, if you click through to [`path-to-regexp`](https://www.npmjs.com/package/path-to-regexp), it mentions that parameters can be optional, i.e. `/path/:param?` should make the type of `req.params.param` be `string | undefined`.

This may be tested elsewhere, but the path can also be a regular expression, and that should continue to work as before (there are no regex literals… yet!).

Reply:
oh, wasn't aware of this. i'll add support for `-`, `.` and `?`, though the regex will probably just have to be ignored

Reply:
How about just adding a test for `/path/:param?` so that we know what the behavior is after this PR? Even if we know it's incorrect, it will be a good place to capture the context from this discussion and what more correct behavior would be in the future.

Reply:
done. though i'm keen to rewrite the `RouteParameters` type to support this as well as properly ignore regex wrapped in parentheses. i think it'll need to be more like a lexer. i'll get on that tomorrow

Reply:
#53513

Reply:
#53513